### PR TITLE
Fix QueryAssertions to use logical types in error message generation

### PR DESCRIPTION
**Pull request by Krishna-Prasad-P-V**
QueryAssertions uses variants to compare results and uses the physical type for the output
error message generation.

This PR enables using the expected result TypePtr for the output error generation.

### CODE REVIEW

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:992]: If expectedType is empty, since expectedRows were empty, it means there are actual rows. We need to pass actualRows here.
Can you add a test for this?

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:1048]: We need to pass expectedType as is.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:1051]: This should be nullptr if actualRows is empty.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:1053]: `expType` -> `expectedType`
`actType` -> `actualType`

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/QueryAssertionsTest.cpp:380]: Added.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp:1048]: Fixed.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp:1051]: Fixed.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp:1053]: Fixed.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp:992]: We are already passing actualRows here, right?
```
    ADD_FAILURE() << generateUserFriendlyDiff(
                         expectedRows, actualRows, expectedType)
```

Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp:992]: Fixed. Passed actualType when expectedRows = 0 and actualRows > 0.

**Comment by Krishna-Prasad-P-V** [comment]
@majetideepak - Kindly re-review the changes. All the tests are passing locally. Thanks.

**Comment by Krishna-Prasad-P-V** [comment]
@majetideepak - Kindly re-review the changes. Fixed all comments. Thanks.

**Comment by Krishna-Prasad-P-V** [comment]
@majetideepak - Kindly re-review the changes. Fixed all comments. Thanks.

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/QueryAssertionsTest.cpp:374]: Fixed.

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/QueryAssertionsTest.cpp:368]: Fixed.

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/QueryAssertionsTest.cpp:382]: Fixed.

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/QueryAssertionsTest.cpp:402]: Fixed.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp:991]: Fixed.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp:1014]: Fixed.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp:1028]: Fixed.

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/QueryAssertionsTest.cpp:381]: Fixed. Used Decimal(6,2) for creating the expected/actual rows.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp:1002]: We already have some - `noFloatColumn`, `singleFloatColumn`, `nullDecimalValue`. Do we need a separate one?

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp:1161]: In `variant::toJson(TypePtr& type)` function, we are looking for the `kind_`. For example, the `kind_ == BIGINT` for a row, but the type passed to this function from QueryAssertions will be `type == ROW` as it is RowVectorPtr type. This causes a mismatch is the `switch-case` as conditions don't meet and the outputs are wrong.

Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp:1161]: Attaching a screenshot from debugger for reference. Looking at the left-hand side of this screenshot might give much clarity. Not sure if we should change something here.
Screenshot 2023-08-21 at 9 22 20 PM

Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp:1161]: @majetideepak - Should we write code in `variant::toJson(TypePtr& type)` to compare the `kind_` and `type` and do the following steps in here? If there isn't much impact, I can do that. Not sure if it will impact some other modules here.

Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp]: Fixed. Added some new code in `variant::toJson()` to handle this.

File: velox/type/Variant.cpp
Comment (majetideepak) [inline velox/type/Variant.cpp:204]: `type->kind()` should do here?
add a check `VELOX_CHECK_EQ(type->kind(), kind_, "Type kind mismatch {} != {}",type->kind(), kind_);`

Comment (Krishna-Prasad-P-V) [reply velox/type/Variant.cpp:204]: I found the issue. We can't use `expected.at(0)->childAt(0)->type()` to get the expectedType. In various cases, we have expected to have 3 different variant rows, i.e, 3 different children and so picking the first one only is not the right solution. We need to change this logic here.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:936]: Should be `expected.at(0)->type()`

File: velox/type/Variant.cpp
Comment (majetideepak) [reply velox/type/Variant.cpp:204]: It definitely should be should be `expected.at(0)->type()`. You need to include your recursive toJson fix as well.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp:936]: Fixed.

File: velox/type/Variant.cpp
Comment (Krishna-Prasad-P-V) [reply velox/type/Variant.cpp:204]: Fixed.

**Comment by Krishna-Prasad-P-V** [comment]
@majetideepak - Kindly re-review. Thanks.

File: velox/type/Variant.h
Comment (majetideepak) [inline velox/type/Variant.h:401]: nit: keep `///` as this in a header class function.

**Review by majetideepak — APPROVED** [review]
@Krishna-Prasad-P-V this looks good now. Thanks!

**Comment by majetideepak** [comment]
@Krishna-Prasad-P-V I suspect your branch name `KP/queryAssertions_fix` to be causing the CI test issues. Can you simply create another branch `queryAssertions_fix` and open a new PR?

File: velox/type/Variant.h
Comment (Krishna-Prasad-P-V) [reply velox/type/Variant.h]: Fixed.

**Comment by Krishna-Prasad-P-V** [comment]
> @Krishna-Prasad-P-V I suspect your branch name `KP/queryAssertions_fix` to be causing the CI test issues. Can you simply create another branch `queryAssertions_fix` and open a new PR?

@majetideepak - It looks like I need a maintainer approval to start the workflow. It says like that.
Screenshot 2023-09-01 at 9 33 07 AM

**Comment by Krishna-Prasad-P-V** [comment]
@mbasmanova - Can I get a maintainer approval for this PR as the pipelines are not running? Thanks.

**Comment by majetideepak** [comment]
@Krishna-Prasad-P-V the maintainer approval is for the python jobs. Not for the regular jobs. Can you please create a new branch as suggested above?

**Comment by majetideepak** [comment]
I just approved the maintainer-run jobs. It did not launch the regular CI jobs.

**Comment by majetideepak** [comment]
I see that you created a branch here https://github.com/facebookincubator/velox/pull/6386 and that did not help.

File: velox/type/Variant.cpp
Comment (mbasmanova) [inline velox/type/Variant.cpp:193]: No need to include type kinds in the message. These are included by default.

File: velox/type/Variant.cpp
Comment (mbasmanova) [inline velox/type/Variant.cpp:226]: same

File: velox/exec/tests/utils/QueryAssertions.h
Comment (mbasmanova) [inline velox/exec/tests/utils/QueryAssertions.h:221]: RowType & expectedRowType

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (mbasmanova) [inline velox/exec/tests/QueryAssertionsTest.cpp:378]: This error message is difficult to read. Would you split it into separate strings, one string per line?

File: velox/exec/tests/utils/QueryAssertions.h
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.h:221]: Did you mean the change as `const RowTypePtr& expectedRowType`?

Comment (mbasmanova) [reply velox/exec/tests/utils/QueryAssertions.h:221]: That's right.

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (majetideepak) [inline velox/exec/tests/QueryAssertionsTest.cpp:393]: @Krishna-Prasad-P-V I don't see the logical type value `40.00` here. Can you verify?

File: velox/type/Variant.h
Comment (majetideepak) [reply velox/type/Variant.h:399]: @mbasmanova With this change, all `variant::toJson()` calls are invoked using a valid TyprPtr. The default `nullptr` can be removed. This will avoid additional checks and handling for `type == nullpr` inside toJson().

File: velox/core/tests/TypedExprSerdeTest.cpp
Comment (majetideepak) [reply velox/core/tests/TypedExprSerdeTest.cpp:67]: `127LL` should work as well. `127` is being mapped to a variant with `typeKind INTEGER`. However, the type specified was BIGINT().

Comment (majetideepak) [reply velox/core/tests/TypedExprSerdeTest.cpp:67]: We should probably add a check inside `ConstantTypedExpr` constructor to avoid this.

File: velox/type/Variant.h
Comment (mbasmanova) [reply velox/type/Variant.h:399]: That sounds good, but doesn't match the PR title / description. Let's make this change in a separate PR.

File: velox/exec/tests/utils/QueryAssertions.h
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.h:221]: Fixed.

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (mbasmanova) [reply velox/exec/tests/QueryAssertionsTest.cpp:378]: Maybe, put each line in a separate string and move "\n" to the end:

```
EXPECT_NONFATAL_FAILURE(
      assertQuery(plan, "SELECT c0 + 1 FROM tmp"),
      "Expected 2, got 2\n"
      "2 extra rows, 2 missing rows\n"
      ...);
```

File: velox/type/Variant.h
Comment (mbasmanova) [reply velox/type/Variant.h:399]: It would be best to limit each PR to one fix.

File: velox/type/Variant.cpp
Comment (Krishna-Prasad-P-V) [reply velox/type/Variant.cpp:226]: Fixed.

File: velox/type/Variant.cpp
Comment (Krishna-Prasad-P-V) [reply velox/type/Variant.cpp:193]: Fixed.

File: velox/type/Variant.h
Comment (Krishna-Prasad-P-V) [reply velox/type/Variant.h:399]: Fixed. Will raise another PR.

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/QueryAssertionsTest.cpp:378]: Fixed. Changed the same for all new tests implemented.

File: velox/type/Variant.h
Comment (majetideepak) [reply velox/type/Variant.h:399]: @Krishna-Prasad-P-V let's move `toJson()` recursive fix separately by using 
```
  friend std::ostream& operator<<(std::ostream& stream, const variant& k) {
    const auto type = inferType();
    stream << k.toJson(type);
    return stream;
  }
```

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/QueryAssertionsTest.cpp:393]: Fixed. Look for line 991 in QueryAssertions.cpp for the change done.

File: velox/type/Variant.cpp
Comment (majetideepak) [inline velox/type/Variant.cpp:207]: this can remain `kind_`

Comment (Krishna-Prasad-P-V) [reply velox/type/Variant.cpp]: Fixed.

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (majetideepak) [reply velox/exec/tests/QueryAssertionsTest.cpp:378]: @Krishna-Prasad-P-V this change seems to be missing. Can you check?

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:1083]: Why do we need these changes?

**Review by majetideepak — COMMENTED** [review]
@Krishna-Prasad-P-V the build is failing. Please take a look.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp]: Rebase issue. Fixed.

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/QueryAssertionsTest.cpp]: Rebase issue. Fixed.

File: velox/exec/tests/QueryAssertionsTest.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/QueryAssertionsTest.cpp]: Fixed.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:1105]: We can always pass the `expectedType` and `actualType` as is here.

Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp]: Fixed.

Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp]: `actualType` cannot be passed as such as the `results` vector for an empty resultset could be 0.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:1083]: Why do we have the materialize the `actual` input here? Can't we simply use
```
const auto actualType =
      (actual.size() == 0) ? nullptr : actual.at(0)->type();
```

Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp]: Right, we don't need materialize here. Removed the same.

**Review by majetideepak — APPROVED** [review]

**Comment by majetideepak** [comment]
@mbasmanova, @kgpai can you take another look at this PR? Thanks.

**Comment by mbasmanova** [comment]
Seeing lint warning:

Screenshot 2023-10-24 at 10 26 00 AM

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:1019]: `const auto& type`

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:1082]: `const auto& actualType`

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:1104]: const auto& actualType

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:1423]: `const auto& type`

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:962]: const auto& expectedType

Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp]: Fixed.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp]: Fixed.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp]: Fixed.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp]: Fixed.

File: velox/exec/tests/utils/QueryAssertions.cpp
Comment (Krishna-Prasad-P-V) [reply velox/exec/tests/utils/QueryAssertions.cpp]: Fixed.

**Pull request closed without merging**
